sw/host/opentitanlib/src/bootstrap/rescue.rs

```diff
@@ -0,0 +1,352 @@
+                },
+                ..Default::default()
+            };
+            let slice_size = Self::DATA_LEN.min(payload.len() - addr);
+            frame.data[..slice_size].copy_from_slice(&payload[addr..addr + slice_size]);
+            frames.push(frame);
+
+            addr += Self::DATA_LEN;
+            frame_num += 1;
+        }
+        frames.last_mut().map(|f| f.header.frame_num |= Self::EOF);
+        frames
+            .iter_mut()
+            .for_each(|f| f.header.hash = f.header_hash());
+        Ok(frames)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum RescueError {
+    #[error("Unrecognized image file format")]
+    ImageFormatError,
+    #[error("Synchronization error communicating with boot rom")]
+    SyncError,
+    #[error("Repeated errors communicating with boot rom")]
+    RepeatedErrors,
+}
+
+/// Implements the UART rescue protocol of Google Ti50 firmware.
+pub struct Rescue {}
+
+impl Rescue {
+    const MAX_CONSECUTIVE_ERRORS: u32 = 50;
+
+    /// Creates a new `Rescue` protocol updater from `options`.
+    pub fn new(_options: &BootstrapOptions) -> Self {
+        Self {}
+    }
+
+    fn read_char(&self, uart: &dyn Uart) -> Result<char> {
+        let mut buf = [0u8; 1];
+        match uart.read_timeout(&mut buf, Duration::from_millis(100)) {
+            Ok(1) => Ok(buf[0] as char),
+            Ok(_) => Err(anyhow!("")),
+            _ => Err(anyhow!("")),
+        }
+    }
+
+    fn expect_string(&self, uart: &dyn Uart, s: &str) -> Result<()> {
+        let char_vec: Vec<char> = s.chars().collect();
+        let mut char_vec: &[char] = &char_vec;
+        while char_vec.len() > 0 {
+            if self.read_char(uart)? != char_vec[0] {
+                return Err(anyhow!(""));
+            }
+            char_vec = &char_vec[1..];
+        }
+        Ok(())
+    }
+
+    fn flush_rx(&self, uart: &dyn Uart) {
+        let mut response = [0u8; Frame::HASH_LEN];
+        loop {
+            match uart.read_timeout(&mut response, Duration::from_millis(500)) {
+                Ok(0) => break,
+                Ok(_) => continue,
+                Err(_) => break,
+            }
+        }
+    }
+
+    /// As the 1024 byte blocks sent to the chip have no discernible header, the sender and
+    /// receiver could be "out of sync".  This is resolved by sending one byte at a time, and
+    /// observing when the chip sends a response (which will be a rejection due to checksum).
+    fn synchronize(&self, uart: &dyn Uart) -> Result<()> {
+        // Most likely, only a few "extra" bytes have been sent during initial negotiation.
+        // Send almost a complete block in one go, and then send each of the last 16 bytes one
+        // at a time, slowly enough to detect a response before sending the next byte.
+        uart.write(&[0u8; 1008])?;
+        let mut response = [0u8; 1];
+        let limit = match uart.read_timeout(&mut response, Duration::from_millis(50)) {
+            Ok(0) => 16,
```

Comment:
You can merge these arms using `Ok(0) | Err(_)`.

Reply:
I have done so here, and above in flush_rx().